paraview/vtk_gcl_converter/vtk_gcl_converter.1
.TH VTK_GCL_CONVERTER 1
.SH NAME
vtk_gcl_converter - converts various primitives for vtk/paraview
.SH SYNOPSIS
.nf
vtk_gcl_converter [ [-lines|polygons|vectors] -noz -clip -pf pffile] 
.fi
.SH DESCRIPTION
.LP
This is a unix filter that can be used to convert geographic data of 
several common types to 3D graphic primitives that can be displayed
with paraview (www.paraview.org).   
Accepted primitives are:  (1) points (the default), (2) lines, 
(3) polygons, and (4) vectors. 
Expected input varies with the
type of objects to be drawn.  Output is always an ascii format
"legacy vtk file" that should be given the file extension ".vtk".  
.LP
Techniques to view the results also are variable.  
\fIPoints\fR are not viewable in paraview, but the file output must be 
passed through the Glphy paraview filter.  Experience has shown that small 
radius spheres a useful primitive to view points.  \fIlines\fR and 
\fIpolygons\fR display directly, but commonly require color changes to 
complement other elements of a 3D scene.  \fIvectors\fR like points need
to be passed through the Glyph filter to be visible.   However, in the
vector case the current default glyph is a vector that can be displayed 
well provided a reasonable scaling is used to make the vectors visible.
.LP
In all cases the input data is geographic and converted for viewing by
paraview to a local coordinate system defined by the parameter file
(see below for details).  
.SH OPTIONS
.LP
First note the -lines, -vectors, or -polygon switch.   If none of these
appear the default of points is used.  If more than one appear the last one 
in the argument list will be used.  
.IP -noz
If this flag appears the depth coordinate is set to zero for all input.
This flag is only used for points or lines.   I is simply ignored for
other inputs.
.IP -clip
This is a boolean switch.  If this flag is given the output will
be clipped in a box defined by the
parameter file.  This option is not available for polygons because
of the complexity of truncating polygons at the edge of a box.   For 
the -line option the curves will be truncated at the clip edge not 
extended to the edge of the box.  
This option is not available for polygons because
of the complexity of truncating polygons at the edge of a box.   For 
the -line option the curves will be truncated at the clip edge not 
extended to the edge of the box.  
.IP -pf
Use pffile instead of the default of vtk_gcl_converter.pf.   Very 
important warning about this program is that it does not provide
the normal antelope pf feature of a path search but the pf file
must be in the current directory.   This was done to allow this
program to be used outside of antelope.
.SH FILES
.LP
The input is always assumed to be a simple ascii file with columns
of numbers.   Files can have comments defined as having a # sign in 
column 1 of the file.   (Warning:  no leading blank characters 
should appear before a # sign as the test used is the character in 
column 1.)  A column header line with a leading # sign is thus strongly
advised for all input files.  Note that in all cases longitude and
latitude input are assumed to be in decimal degrees and depths are 
assumed to be in kilometers.
.LP
\fIPoint\fR data have the simplest format.  Three columns of free field
input in this order:  longitude, latitude, and depth. 
.LP
\fIVector\fR data have added columnes to define the vector to be drawn
at a given position.   The order is:
longitude, latitude, depth, east component of vector, north component
of vector, vertical component of vector.   Note the vector is assumed
in geographic coordinates at the point give.  That is, east, north,
and up positive with north defined by geographic north at that point.
.LP
\fILines\fR and \fIpolygons\fR assume a gmt style line drawing format.
That is, the basic file is three columns like points (longitude,
latitude, and depth values).  These points are linked by line segments.
Line breaks are defined by a break sentinel.   This is currently frozen
as the gmt default "<" character.  
.SH PARAMETER FILE
.LP
The most important parameters for this program are the following 
that define the Cartesian coordinate system the data are translated
to for viewing with paraview.  \fIorigin_latitude\fR,
\fIorigin_longitude\fR, and \fIorigin_radius\fR define the origin 
of the coordinate system (latitude and longitude in degrees and
radius in km).  Note that elevation data are measured relative to 
a reference ellipsoid used in gclgrid library so this radius needs
to be consistent with that definition.   One final important
coordinate parameter is \fIazimuth_y_axis\fR, which is the azimuth of
the Cartesian y axis from local north at the origin. This is 
inherited from the coordinate system used in libgclgrid for defining
a regular grid rotated by a specified angle.   
This angle is assumed in degrees.
.LP
The following parameters are used to define the bounding box when
the -clip option is requested:  
\fIx1low, x1high, x2low, x2high, x3low,\fR and \fIx3high\fR.
These define a box in the Cartesian system with clipping outside
the range of the low to high value.  All are assume in units of km.
.LP
The following are used only with the -vector data type:
\fIread_vector_attribure, vector_scale_factor, scalars_tag_amplitude,
vectors_tag,\fR and \fIvectors_tag\fR.  
The most important parameter that may require tweeking is \fIvector_scale_factor\fR.
All input vector data are multiplied by this constant to be drawn at a 
size that will be visible in the given scene.  That constant depends on the
scale of the scene.  Note this is just a base scale factor as the 
Glyph filter in paraview has a scaling parameter that can be used to tweek
the scale factor interactively.   Generally one should choose this to 
make sure the original vectors are at least visible so you aren't confused 
by no visible vectors.
If \fIread_vector_attribure\fR is set true (default is false) the 
scalars attribute that paraview uses to for color mapping will be read
from the input file.  
\fIscalars_tag\fR defines the tag to label the legend for the 
associated color map for this vector attribute.
\Ivectors_tag\fR sets the tag on 
the vectors section of the output vtk file.  At this writing I do not
know how this could be used as it seems like pure baggage for now. Made
it a variable as there is likely some use I don't know about.
.SH EXAMPLE
.LP
To convert a set polygon data in file polys.dat.
.nf
vtk_gcl_converter -polygons < polys.dat > polys.vtk
.fi
Note that the .vtk extension is used by paraview to recognize this as
a "legacy vtk file".
.LP
A less than obvious example of the use of this program is
to draw pscoast data that can be obtained from gmt using 
forms similar to this:
.nf
pscoast -R-94/-80/32/42 -JM5.0 -Df -m | vtk_gcl_converter -lines > coast.vtk
.fi
where the critical parameters are the region (defined by -R) and the 
-m flag that caues pscoast to spit out coordinates instead of a postscript
graphic (see GMT documentation for details).   
.SH DIAGNOSTICS
.LP
Various "MetadataGetError" messages can appear for parameter file 
deficiences.
.SH "SEE ALSO"
.nf
makegclgrid(1)
.fi
.SH "BUGS AND CAVEATS"
.LP
This program has evolved with feature creep.  It started with 
points and line and everything else is version creep.   
There are thus a variety of inconsistencies.   
.LP
Because the program is a unix filter just typing the command name 
can produce confusing output.  Use a --help or any other random illegal
argument to get the usage line.  That is not the ideal way to do that.
.SH AUTHOR
.nf
Gary L. Pavlis
Department of Geological Sciences
Indiana University
pavlis@indiana.edu
.fi


paraview/vtk_gcl_converter/vtk_gcl_converter.cc
#include <iostream>
#include <vector>
#include <list>
#include "seispp.h"
#include "gclgrid.h"
#include "RegionalCoordinates.h"
#include "PfStyleMetadata.h"
class BoundingBox
{
public:
    BoundingBox(double x1l,double x1h,double x2l,double x2h,
            double x3l,double x3h);
    bool inside(Cartesian_point cp);
private:
    double x1low,x1high;
    double x2low,x2high;
    double x3low,x3high;
};
BoundingBox::BoundingBox(double x1l, double x1h,
        double x2l,double x2h,double x3l,double x3h)
{
    x1low=x1l;
    x1high=x1h;
    x2low=x2l;
    x2high=x2h;
    x3low=x3l;
    x3high=x3h;
}
bool BoundingBox::inside(Cartesian_point cp)
{
	if(cp.x1>x1high) return(false);
	if(cp.x1<x1low) return(false);
	if(cp.x2>x2high) return(false);
	if(cp.x2<x2low) return(false);
	if(cp.x3>x3high) return(false);
	if(cp.x3<x3low) return(false);
	return(true);
}

using namespace std;
using namespace SEISPP;
enum ObjectType{POINTS,LINES,POLYGONS,VECTORS};
void usage()
{
	cerr << "vtk_gcl_converter [ [-lines|polygons|vectors] -noz -clip -pf pffile] < infile "<<endl
		<< "Converted data written to stdout.  Default format points"<<endl;
	exit(-1);
}

bool SEISPP::SEISPP_verbose(false);
int main(int argc, char **argv)
{
    ObjectType objtype(POINTS);
    bool noz(false);
    bool clip(false);
    int i;
    string pffile("vtk_gcl_converter.pf");
    for(i=1;i<argc;++i)
    {
	string testarg(argv[i]);
	if(testarg=="-lines")
		objtype=LINES;
        else if(testarg=="-polygons")
                objtype=POLYGONS;
        else if(testarg=="-vectors")
                objtype=VECTORS;
	else if(testarg=="-noz")
		noz=true;
	else if(testarg=="-clip")
		clip=true;
	else if(testarg=="-pf")
        {
            ++i;
            if(i>=argc) usage();
            pffile=string(argv[i]);
        }
	else
	    usage();
    }
    try {
        PfStyleMetadata control
                    =pfread(pffile);
        double olat,olon,oradius,azimuth_y;
        olat=control.get_double("origin_latitude");
        olon=control.get_double("origin_longitude");
        oradius=control.get_double("origin_radius");
        azimuth_y=control.get_double("azimuth_y_axis");
        olat=rad(olat);
        olon=rad(olon);
        azimuth_y=rad(azimuth_y);
        RegionalCoordinates coords(olat,olon,oradius,azimuth_y);
        /* Construct a bounding box from parameter file data */
        double x1l=control.get_double("x1low");
        double x1h=control.get_double("x1high");
        double x2l=control.get_double("x2low");
        double x2h=control.get_double("x2high");
        double x3l=control.get_double("x3low");
        double x3h=control.get_double("x3high");
        BoundingBox bb(x1l,x1h,x2l,x2h,x3l,x3h);
        double vector_scale_factor(1.0);
        string scalars_tag,vectors_tag;
        bool read_vector_attribute;
        if(objtype==VECTORS)
        {
            vector_scale_factor=control.get_double("vector_scale_factor");
            scalars_tag=control.get_string("scalars_tag");
            vectors_tag=control.get_string("vectors_tag");
            read_vector_attribute=control.get_bool("read_vector_attribute");
        }
	vector<Cartesian_point> cplist;
	vector<Geographic_point> gplist;
	Geographic_point gp;
	Cartesian_point cp;
	double lat,lon,depth,radius;
		/* First write common data for either points or lines for file head */
	cout << "# vtk DataFile Version 2.0"<<endl
          << argv[0] << " conversion"  <<endl
          << "ASCII" <<endl;
        if(objtype==VECTORS)
            cout << "DATASET UNSTRUCTURED_GRID"<<endl;
        else
            cout << "DATASET POLYDATA"<<endl;
		typedef list<int> LineLinks;
		LineLinks thissegment;
		list<LineLinks> segments;
        list<double> east,north,vertical,attributes;  //for VECTORS
		char line[256];
		int i=0;
        switch(objtype)
        {
        case LINES:
        case POLYGONS:
            while(!cin.eof())
    	    {
        	    cin.getline(line,256);
        	    if(cin.eof())
        	    {
        		/*this is an odd logic construct, but
        		we make this look like a > line to force
        		the last segment to be pushed to the segments
        		list.  We make sure the size isn't zero */
                    if(thissegment.size()>1)
        		{
        			line[0]='>';
        			line[1]='\0';
        		}
        	    }
        	    if(line[0]=='#') continue;
        	    if(line[0]=='>')
        	    {
        		if(thissegment.size()>1)
        		{
        			segments.push_back(thissegment);
        			thissegment.clear();
        		}
        		else if(thissegment.size()==1)
        		{
        		/* Handle special case possible when clipping.
        				This is if a line wanders in and out of the bounding
        				box it is possible to end up with a stray point we
        				need to clear like this. */
        			thissegment.clear();
        		}
        	    }
        	    else
        	    {
        		if(noz)
        		{
        		    sscanf(line,"%lf%lf",&lon,&lat);
        		    depth=0.0;  // inefficient but clearer
        		}
        		else
        		{
        			sscanf(line,"%lf%lf%lf",&lon,&lat,&depth);
        		}
        		gp.lat=rad(lat);
        		gp.lon=rad(lon);
        		gp.r=r0_ellipse(gp.lat)-depth;
                        cp=coords.cartesian(gp);
        		if(clip)
        		{
        		    if(bb.inside(cp))
        		    {
        			cplist.push_back(cp);
        			thissegment.push_back(i);
        			++i;
        		    }
        		    else
        		    {
        			/* truncate segment if it wanders outside box.
                                   Drop if there is only one point. */
        			if(thissegment.size()>1)
        				   segments.push_back(thissegment);
        			thissegment.clear();
        		    }
        		}
        		else
        		{
        		    cplist.push_back(cp);
        		    thissegment.push_back(i);
        		    ++i;
        		}
                }
            } 
            break;
        case VECTORS:
	    while(cin.getline(line,256))
	    {
                double e,n,z;
                double magvec;
                double vector_attribute;
                stringstream ss(line);
                /*
                sscanf(line,"%lf%lf%lf%lf%lf%lf",
                                &lat,&lon,&depth,&e,&n,&z);
                                */
                ss >> lon; ss >> lat; ss >> depth;
                  ss>>e;  ss>>n;  ss>>z;
                magvec=sqrt(e*e+n*n+z*z);
                if(read_vector_attribute) 
                    ss >>vector_attribute;
                else
                    /* Default is amplitude */
                    vector_attribute=magvec;
		gp.lat=rad(lat);
		gp.lon=rad(lon);
		gp.r=r0_ellipse(gp.lat)-depth;
                        cp=coords.cartesian(gp);
		if(clip)
		{
		    if(bb.inside(cp))
                    {
			cplist.push_back(cp);
                        east.push_back(e);
                        north.push_back(n);
                        vertical.push_back(z);
                        gplist.push_back(gp);
                        attributes.push_back(vector_attribute);
                    }
		}
		else
                {
		    cplist.push_back(cp);
                    east.push_back(e);
                    north.push_back(n);
                    vertical.push_back(z);
                    gplist.push_back(gp);
                    attributes.push_back(vector_attribute);
                }
            }
            break;
        case POINTS:
        default:
	    while(cin.getline(line,256))
	    {
	    	if(noz)
		{
			sscanf(line,"%lf%lf",&lon,&lat);
			depth=0.0;
		}
		else
		{
			sscanf(line,"%lf%lf%lf",&lon,&lat,&depth);
		}
		gp.lat=rad(lat);
		gp.lon=rad(lon);
		gp.r=r0_ellipse(gp.lat)-depth;
                        cp=coords.cartesian(gp);
		if(clip)
		{
			if(bb.inside(cp))
				cplist.push_back(cp);
		}
		else
			cplist.push_back(cp);
            }
	};
	cout << "POINTS " << cplist.size() << " float"<<endl;
	vector<Cartesian_point>::iterator cpptr;
        vector<Geographic_point>::iterator gpptr;
	for(cpptr=cplist.begin();cpptr!=cplist.end();++cpptr)
	{
	    cout << cpptr->x1 << " "
			<< cpptr->x2<<" "
			<< cpptr->x3<<" "<<endl;
	}
	int sizepar;
	list<LineLinks>::iterator llptr;
	LineLinks::iterator lineptr;
        list<double>::iterator ie,in,iz,ia;
        int nv;
        switch(objtype)
	{
        case VECTORS:
            nv=east.size();   //use e assuming all same
            /* First write the scalars block */
            cout << "POINT_DATA "<<nv<<endl;
            cout << "SCALARS "<<scalars_tag<<" float 1"<<endl;
            cout << "LOOKUP_TABLE default"<<endl;
            for(ia=attributes.begin();ia!=attributes.end();++ia)
                        cout << *ia <<endl;
             /* Last thing is to write the vectors block.  These
                       we scale by the factor given in the parameter file.
                       Note amps values are intentionally NOT scaled by
                       this factor so they retain original vector units. */
            cout << "VECTORS "<<vectors_tag<<" float"<<endl;
            for(ie=east.begin(),in=north.begin(),
                        iz=vertical.begin(),gpptr=gplist.begin();
                        ie!=east.end();++ie,++in,++iz,++gpptr)
            {
                    dmatrix l2r=coords.l2rtransformation(gpptr->lat,
                            gpptr->lon);
                    dvector v(3);
                    v(0)=(*ie);
                    v(1)=(*in);
                    v(2)=(*iz);
                    v=l2r*v;
                    /* dvector at time this written did not support
                       multiplication by scale like dmatrix does so
                       need this loop */
                    for(i=0;i<3;++i)
                        v(i)*=vector_scale_factor;
                    cout << v(0)<<" "<<v(1)<<" "<<v(2)<<endl;
            }
            break;
        case POLYGONS:
        case LINES:
        /* For this stupid format we have to count the total number
		of items in segments and number of points for the "size" parameter
		of the LINES keyword line. */
		sizepar=0;
		for(llptr=segments.begin();llptr!=segments.end();++llptr)
		{
		    for(lineptr=llptr->begin();lineptr!=llptr->end();++lineptr)
					++sizepar;
		}
		sizepar+=segments.size();
                switch(objtype)
                {
                    case POLYGONS:
                        cout << "POLYGONS ";
                        break;
                    case LINES:
                    default:
		        cout <<  "LINES "; 
                }
                cout << segments.size() << " " <<sizepar<<endl;
		for(llptr=segments.begin();llptr!=segments.end();++llptr)
		{
			cout << llptr->size() << " ";
			for(lineptr=llptr->begin();lineptr!=llptr->end();++lineptr)
		        {
					cout << *lineptr<<" ";
			}
			cout << endl;
		}
                break;
        default:
                cout <<endl;
        };
    } catch (SeisppError& serr)
    {
	serr.log_error();
	exit(-1);
    }
    catch(std::exception& stdexcp)
    {
        cerr << stdexcp.what()<<endl;
    }
}
